### PR TITLE
Expose provider tracker to service factory

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following exposes the provider tracker to the service
factory. The service factory now takes a IAAS and CAAS provider
tracker. Initially it was hoped that you could use the generic
provider tracker directly, but for controllers with both IAAS
and CAAS mode types this isn't possible. We do need to expose both
at the service factory level.

The service that requires a provider can then create a local interface
that implements some functionality of an environ. For instance the
network domain service could expose that it would like a provider
that supports networking:

    type Provider interface {
        environs.Networking
    }

The service then takes a `providertracker.ProviderGetter[Provider]`,
which will hand out providers. If the service requires to work on
both IAAS and CAAS, then two provider getters will needed as dependencies.

To provide the provider getter to the service, the service factory
needs to supply the right provider.

    providertracker.ProviderRunner[networkservice.Provider](s.providerFactory, s.modelUUID.String())

Using a provider runner we can late bind the provider for the model
UUID, so the error surfaces on use of the service method. This is
akin to the DB pattern we currently have.

This ProviderRunner replaces the SupportsNetworking and other
environs casting functions. Ideally the casting functions in the
environs package can be removed.

The only concern with this change, is that the core/providertracker
package is bringing in the environs package. It's clear that it
_should_ live there, it just breaks one of the rules of the core
package.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test
$ juju add-model test
$ juju deploy ubuntu
```

## Links

**Jira card:** JUJU-5644

### CODE REVIEW

File: domain/network/service/space.go [hunk 1]
```diff
@@ -136,22 +138,21 @@ func (s *Service) SaveProviderSubnets(
 				continue
 			}
 			fanSubnetID := generateFanSubnetID(subnetNet.String(), subnet.ProviderId.String())
-			if overlaySegment != nil {
```
Comment (SimonRichardson) [inline domain/network/service/space.go:139]: This was a tautological lint error. We already check for `overlaySegment == nil` just above the continue in the diff.

File: cmd/jujud-controller/agent/machine/manifolds.go [hunk 2]
```diff
@@ -865,6 +873,30 @@ func commonManifolds(config ManifoldsConfig) dependency.Manifolds {
 			GetControllerConfigService: objectstores3caller.GetControllerConfigService,
 			NewWorker:                  objectstores3caller.NewWorker,
 		})),
+
+		caasProviderTrackerName: ifDatabaseUpgradeComplete(providertracker.MultiTrackerManifold[caas.Broker](providertracker.ManifoldConfig[caas.Broker]{
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/jujud-controller/agent/machine/manifolds.go:877]: We upgrade the database (state and domain) in order to use the trackers. The database upgrade must be completed before being allowed to use them.

File: domain/servicefactory/doc.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+//                            └──────┬──────┘
+//                                   │
+//                                   │
+//                            ┌──────▼──────┐
+//                            │             │
+//                            │             │
+//                            │   OUTPUT    │
+//                            │             │
+//                            │             │
+//                            └─────────────┘
```
Comment (nvinuesa) [inline domain/servicefactory/doc.go:81]: amazing! :rocket:

File: domain/network/service/space.go [hunk 4]
```diff
@@ -169,3 +170,26 @@ func generateFanSubnetID(subnetNetwork, providerID string) string {
...
+}
+
+// NewProviderService returns a new service reference wrapping the input state.
+func NewProviderService(st State, provider providertracker.ProviderGetter[Provider], logger Logger) *ProviderService {
+	return &ProviderService{
+		Service: Service{
+			st:     st,
+			logger: logger,
+		},
+		provider: provider,
```
Comment (nvinuesa) [inline domain/network/service/space.go:193]: This means that now we'll have the service layer split between `Service` and `ProviderService`. Can't we have the provider as part of the `Service` (with a check for nil if used in a method)? that way we have everything under the same struct

Comment (SimonRichardson) [reply domain/network/service/space.go:193]: Correct. This is on purpose. We have the same split between `Service` and `WatchableService`. This is to give guard rails for things like model migrations, which don't require watchers. I want people to think __really__ hard when adding model migrations to understand if they also need a provider. I hope they don't, it should reduce the complexity of model migrations and reduce the exposed testing. If they do require a provider then so be it, but it's a conversation they've at least had with themselves.

File: core/providertracker/providertracker.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	"context"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/environs"
+)
+
+// Provider in the intersection of a environs.Environ and a caas.Broker.
+//
+// We ideally don't want to import the environs package here, but I've not
```
Comment (manadart) [inline core/providertracker/providertracker.go:16]: Worth adding comment detail to the effect of not being useful in and of itself, but serves as the constraint for generic parameterisation, preferable to `any`.

File: domain/servicefactory/doc.go (hunk 3 above)
Comment (manadart) [reply domain/servicefactory/doc.go:81]: Yep. Good doc.

File: internal/worker/providertracker/manifold.go [hunk 6]
```diff
@@ -270,40 +224,165 @@ func manifoldOutput[T Provider](in worker.Worker, out any) error {
...
+func (f serviceFactory) Config() ConfigService {
+	return f.factory.Config()
+}
+
+// Credential returns the provider credential service.
+func (f serviceFactory) Credential() CredentialService {
+	return f.factory.Credential()
+}
+
+// GetModelProviderServiceFactoryGetter is a helper function that gets a service
```
Comment (manadart) [inline internal/worker/providertracker/manifold.go:335]: It doesn't get a service though, does it?

File: rpc/client.go [hunk 7]
```diff
@@ -201,6 +201,14 @@ func (call *Call) done() {
 // The params value may be nil if no parameters are provided; the response value
 // may be nil to indicate that any result should be discarded.
 func (conn *Conn) Call(ctx context.Context, req Request, params, response interface{}) error {
+	// Before sending the request, check if the context has been canceled.
+	// This is done to prevent any unnecessary work from being done if the
```
Comment (manadart) [inline rpc/client.go:205]: "if it has." reads better.

**Review by manadart — APPROVED** [review]
This is the pointy end of some truly marvelous work. My nits are only regarding comments, so we're good to go here. I will land it.

**Pull request merged**